### PR TITLE
VRF: move from `cloud` APIs to `evpn-gw` APIs

### PR DESCRIPTION
**Pull request by glimchb**
- Note the collapse of `CreateTunnel` and `CreateSubnet` into `CreateVrf`
- Note RouteTable and BridgeMac are randomly generated now

Signed-off-by: Boris Glimcher <Boris.Glimcher@emc.com>

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	"net"
 
 	pb "github.com/opiproject/opi-api/network/cloud/v1alpha1/gen/go"
```
Comment (mardim91) [inline cmd/main.go:14]: This needs to be removed right ?

File: cmd/main.go [hunk 2]
```diff
@@ -33,6 +34,7 @@ func main() {
 
 	// TODO: replace cloud -> evpn
 	pb.RegisterCloudInfraServiceServer(s, opi)
```
Comment (mardim91) [inline cmd/main.go:36]: I guess also this one needs to be removed

File: README.md [hunk 3]
```diff
@@ -30,18 +30,18 @@ Run `docker-compose up -d`
 ## Manual gRPC example
 
 ```bash
-docker-compose exec opi-evpn-bridge grpcurl -plaintext -d '{"vpc" : {"spec" : {"v4_route_table_name_ref" : "1234"} }, "vpc_id" : "testvpc", "parent" : "todo" }' localhost:50151 opi_api.network.cloud.v1alpha1.CloudInfraService.CreateVpc
-docker-compose exec opi-evpn-bridge grpcurl -plaintext -d '{"subnet" : {"spec" : {"vpc_name_ref": "//network.opiproject.org/vpcs/blue", "virtual_router_mac": "qrvMAAAB", "v4_prefix": {"addr": 336860161, "len": 24} } }, "subnet_id" : "testbridge", "parent" : "todo" }' localhost:50151 opi_api.network.cloud.v1alpha1.CloudInfraService.CreateSubnet
+docker-compose exec opi-evpn-bridge grpcurl -plaintext -d '{"vrf" : {"spec" : {"vni" : 1234, "loopback_ip_prefix" : {} }}, "vrf_id" : "testvrf" }' localhost:50151 opi_api.network.cloud.v1alpha1.CloudInfraService.CreateVrf"
```
Comment (mardim91) [inline README.md:33]: Why the loopback_ip_prefix is empty ?

File: docker-compose.yml [hunk 4]
```diff
@@ -270,27 +270,30 @@ services:
...
+      /entrypoint.sh call --json_input --json_output localhost:50151 CreateVrf    "{\"vrf\" : {\"spec\" : {\"vni\" : 1000, \"loopback_ip_prefix\": {} }}, \"vrf_id\" : \"blue\" }" && \
+      /entrypoint.sh call --json_input --json_output localhost:50151 CreateVrf    "{\"vrf\" : {\"spec\" : {\"vni\" : 1001, \"loopback_ip_prefix\": {} }}, \"vrf_id\" : \"green\" }" && \
+      /entrypoint.sh call --json_input --json_output localhost:50151 CreateVrf    "{\"vrf\" : {\"spec\" : {\"vni\" : 1002, \"loopback_ip_prefix\": {} }}, \"vrf_id\" : \"yellow\" }" && \
+      /entrypoint.sh call --json_input --json_output localhost:50151 CreateSubnet "{\"subnet\" : {\"spec\" : {\"vpc_name_ref\" : \"//network.opiproject.org/vrfs/blue\",   \"virtual_router_mac\": \"qrvMAAAB\" }                                                    }, \"subnet_id\" : \"br100\", \"parent\" : \"todo\" }" && \
```
Comment (mardim91) [inline docker-compose.yml:289]: I guess also the createSubnet needs to be removed and replaced with objects from the EVPN GW API.  If you try to do that one step at the time then is understandable.

File: pkg/evpn/svi_subnet.go [hunk 5]
```diff
@@ -83,8 +83,8 @@ func (s *Server) CreateSubnet(_ context.Context, in *pb.CreateSubnetRequest) (*p
 			log.Printf("error: %v", err)
 			return nil, err
 		}
-		// now get VRF/VPC to plug this bridge into
-		vpc, ok := s.Vpcs[in.Subnet.Spec.VpcNameRef]
+		// now get VRF to plug this bridge into
+		vpc, ok := s.Vrfs[in.Subnet.Spec.VpcNameRef]
```
Comment (mardim91) [inline pkg/evpn/svi_subnet.go:87]: Why a VRF is assigned to a VPC ?

File: pkg/evpn/svi_subnet_test.go [hunk 6]
```diff
@@ -30,7 +30,7 @@ var (
 	testSubnetName = resourceIDToFullName("subnets", testSubnetID)
 	testSubnet     = pb.Subnet{
 		Spec: &pb.SubnetSpec{
-			VpcNameRef:       testVpcName,
+			VpcNameRef:       testVrfName,
```
Comment (mardim91) [inline pkg/evpn/svi_subnet_test.go:33]: Same here.

File: pkg/evpn/vrf.go [hunk 7]
```diff
@@ -26,41 +25,34 @@ import (
...
-		err := resourceid.ValidateUserSettable(in.VpcId)
+	if in.VrfId != "" {
+		err := resourceid.ValidateUserSettable(in.VrfId)
 		if err != nil {
 			log.Printf("error: %v", err)
 			return nil, err
 		}
-		log.Printf("client provided the ID of a resource %v, ignoring the name field %v", in.VpcId, in.Vpc.Name)
-		resourceID = in.VpcId
+		log.Printf("client provided the ID of a resource %v, ignoring the name field %v", in.VrfId, in.Vrf.Name)
```
Comment (mardim91) [inline pkg/evpn/vrf.go:44]: in a creation of an object the VRF.name will be empty initially right ? So there will be no name to print.

File: README.md (hunk 3 above)
Comment (glimchb) [reply README.md:33]: will be filled in following PR, tried to do smallest change possible before complicating implementation

File: cmd/main.go (hunk 2 above)
Comment (glimchb) [reply cmd/main.go:36]: correct, once we complete transition to new APIs, old will be removed

File: docker-compose.yml (hunk 4 above)
Comment (glimchb) [reply docker-compose.yml:289]: correct, one step at a time

File: pkg/evpn/svi_subnet.go (hunk 5 above)
Comment (glimchb) [reply pkg/evpn/svi_subnet.go:87]: variable name is incorrect, will change, thanks for comment

File: pkg/evpn/vrf.go (hunk 7 above)
Comment (glimchb) [reply pkg/evpn/vrf.go:44]: correct

File: cmd/main.go (hunk 1 above)
Comment (glimchb) [reply cmd/main.go:14]: correct, once we complete transition to new APIs, old will be removed

File: pkg/evpn/vrf.go [hunk 8]
```diff
@@ -26,60 +28,116 @@ import (
...
+	vxlanName := fmt.Sprintf("vni%d", in.Vrf.Spec.Vni)
+	myip := make(net.IP, 4)
+	binary.BigEndian.PutUint32(myip, in.Vrf.Spec.VtepIpPrefix.Addr.GetV4Addr())
+	// TODO: take Port from proto instead of hard-coded
+	vxlan := &netlink.Vxlan{LinkAttrs: netlink.LinkAttrs{Name: vxlanName}, VxlanId: int(in.Vrf.Spec.Vni), Port: 4789, Learning: false, SrcAddr: myip}
+	if err := netlink.LinkAdd(vxlan); err != nil {
+		fmt.Printf("Failed to create Vxlan link: %v", err)
+		return nil, err
+	}
+	// Example: ip link set vni100 master br100 addrgenmode none
```
Comment (mardim91) [inline pkg/evpn/vrf.go:121]: ip link set vni100 master br100

This is the command that we have used. Not sure if the difference plays a role.

File: README.md [hunk 9]
```diff
@@ -30,18 +30,18 @@ Run `docker-compose up -d`
 ## Manual gRPC example
 
 ```bash
-docker-compose exec opi-evpn-bridge grpcurl -plaintext -d '{"vpc" : {"spec" : {"v4_route_table_name_ref" : "1234"} }, "vpc_id" : "testvpc", "parent" : "todo" }' localhost:50151 opi_api.network.cloud.v1alpha1.CloudInfraService.CreateVpc
-docker-compose exec opi-evpn-bridge grpcurl -plaintext -d '{"subnet" : {"spec" : {"vpc_name_ref": "//network.opiproject.org/vpcs/blue", "virtual_router_mac": "qrvMAAAB", "v4_prefix": {"addr": 336860161, "len": 24} } }, "subnet_id" : "testbridge", "parent" : "todo" }' localhost:50151 opi_api.network.cloud.v1alpha1.CloudInfraService.CreateSubnet
+docker-compose exec opi-evpn-bridge grpcurl -plaintext -d '{"vrf" : {"spec" : {"vni" : 1234, "loopback_ip_prefix" : {"addr": {"af": "IP_AF_INET", "v4_addr": 167772162} }, "len": 24}, "vtep_ip_prefix": {"addr": {"af": "IP_AF_INET", "v4_addr": 167772162} }, "len": 24} }}, "vrf_id" : "testvrf" }' localhost:50151 opi_api.network.cloud.v1alpha1.CloudInfraService.CreateVrf"
```
[the remark this answers is not in the document]
Comment (venkatmahalingam) [reply README.md:33]: yes, loopback interface IP should be /32, but we need to create a loopback interface and assign the IP, should be assigned  on the VRF interface.

File: pkg/evpn/vrf.go [hunk 10]
```diff
@@ -26,60 +28,116 @@ import (
 	"google.golang.org/protobuf/types/known/emptypb"
 )
 
-// CreateVpc executes the creation of the VRF/VPC
-func (s *Server) CreateVpc(_ context.Context, in *pb.CreateVpcRequest) (*pb.Vpc, error) {
-	log.Printf("CreateVpc: Received from client: %v", in)
+// CreateVrf executes the creation of the VRF
+func (s *Server) CreateVrf(_ context.Context, in *pb.CreateVrfRequest) (*pb.Vrf, error) {
```
[the remark this answers is not in the document]
Comment (venkatmahalingam) [reply pkg/evpn/vrf.go:32]: Sure, we can assign those numbers as per our convenience.

[the remark this answers is not in the document]
Comment (glimchb) [reply pkg/evpn/vrf.go:32]: opened #74 to track this

File: pkg/evpn/vrf.go [hunk 11]
```diff
@@ -26,60 +28,116 @@ import (
...
-	table, err := strconv.Atoi(in.Vpc.Spec.V4RouteTableNameRef)
-	if err != nil {
-		err := status.Error(codes.InvalidArgument, "could not convert V4RouteTableNameRef to integer")
-		log.Printf("error: %v", err)
-		return nil, err
-	}
 	// not found, so create a new one
-	vrf := &netlink.Vrf{LinkAttrs: netlink.LinkAttrs{Name: resourceID}, Table: uint32(table)}
+	vrfName := resourceID
+	tableID := uint32(1000 + math.Mod(float64(in.Vrf.Spec.Vni), 10.0))
```
[the remark this answers is not in the document]
Comment (glimchb) [reply pkg/evpn/vrf.go:59]: opened #74 to track this

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (glimchb) [reply README.md:33]: opened #75 to track this

**Comment by glimchb** [comment]
Merging this PR and opened few issues to address the gaps from the code review

**Pull request merged**